lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -323,9 +323,11 @@ broadenSingleElementStores(StoreInst *storeInst,
   // deletes the instructions or simply records them for later deletion.
   auto nextII = std::next(storeInst->getIterator());
 
-  // Bail if the store's destination is not a struct_element_addr.
+  // Bail if the store's destination is not a struct_element_addr or if the
+  // store's destination (%base) is not a loadable type. If %base is not a
+  // loadable type, we can't create it as a struct later on.
   auto *sea = dyn_cast<StructElementAddrInst>(storeInst->getDest());
-  if (!sea)
+  if (!sea || !sea->getOperand()->getType().isLoadable(*sea->getFunction()))
```

Comment:
That isn't what I mean. I mean assume I have a test case like this:

```
struct Loadable2 { ... }

struct Loadable1 {
   var x: Loadable2
}

struct CPPType {
   var y: Loadable2
}
```

In this case, couldn't a store of loadable 2 be expanded twice?

Reply:
Then you could get SIL like this:

```
 sil shared [transparent] [serializable] [ossa] @test : $@convention(method) (Int32, @thin CPPType.Type) -> @out CPPType {
 bb0(%0 : $*CPPType, %1 : $Int32, %2 : $@thin CPPType.Type):
   %3 = struct_element_addr %0 : $*CPPType, #CPPType.y
   %3a = struct_element_addr %3 : $*Loadable1, #Loadable1.x
   store %1 to [trivial] %3a : $*Int32
   %5 = tuple ()
   return %5 : $()
 }
```

Reply:
Part of the reason why I am pretty sure I am correct is that the unreferenced storage is the same sort of check as what you are trying to do here. In this code both checks should occur together.

Reply:
In fact, I would argue that one in OSSA actually would need to check that it is legal to perform this optimization before creating structs since that can result in those structs being dropped (and thus we will get multiple consumes).

Reply:
I'm going to leave this change here for now (mainly so the tests pass). I've created a more comprehensive fix in #32318. Once that lands, I'll rebase.